assessments/projects/keycloak/README.md

```diff
@@ -0,0 +1,85 @@
+
+
+## Summary
+
+### Design
+
+Keycloak follows a modular and layered approach and is built on top of components such as Java Virtual Machine as a
+trusted runtime environment. It leverages the WildFly application server for REST API implementation, SSL, Data Sources,
+Transaction handling, etc.
+
+Keycloak is intended to be a highly available, scalable and extensible solution providing an implementation for
+modern security standards specifically OAuth2, OIDC.
+
+### Analysis
+
+In today’s microservice-oriented environments which are diverse and ephemeral in nature, there is a need for an
+open-source solution like Keycloak to provide identity and access management features so that applications can deal
+with core functionality instead of dealing with login forms, authenticating users etc.
+
+Keycloak being a critical piece of the infrastructure and its dependency on other components like JVM,
+WildFly Application Server etc. increases its attack surface and could potentially be compromised by attackers.
+Keycloak is a highly customizable solution which is very convenient and powerful for the user, however care should be
+taken to properly understand the consequences of different server configuration options and how administrators define
+fine-grained access to users in different realms to manage the server.
+
+
+## Recommendations
+
+### Recommendations for the project team
+
+* Given Keycloak comprises so many moving parts, it would be helpful to perform threat modelling. [KEYCLOAK-15945](https://issues.redhat.com/browse/KEYCLOAK-15945)
+
+* Keycloak documentation contains great information. However, it could be better organized and consolidated so that
+users can quickly find the right information. The server installation docs in particular could be more concise to help
+users from making any configuration mistakes. It may also be useful to highlight specific terms in the document that
+have security impacting configurations. For example, default configuration is not secure but making change X is
+recommended from a security perspective. In fact, you could even dedicate a section to “Security” in the docs to helps
+users deploy Keycloak in a more secure manner. [KEYCLOAK-15946](https://issues.redhat.com/browse/KEYCLOAK-15946)
+
+* Although it may be useful for the admin to impersonate a user for debugging purposes, it could be leveraged by an
+attacker to take actions like compromising tokens. The impersonated user sessions is a good way to audit any
+illicit activity in the aftermath but its benefits would be restricted to OIDC sessions. Moreover, any user with the
+realm’s impersonation role can impersonate a user. It would be beneficial to make this more explicit in the
+documentation along with the security consequences of this particular setting so that Keycloak adopters can use
+this powerful capability without increasing exposure of their systems to attacks. [KEYCLOAK-15947](https://issues.redhat.com/browse/KEYCLOAK-15947)
```

Comment:
A little verbose, how about something like this (details in issues are fine).

```suggestion
* Keycloak provides the ability for an admin to impersonate a user. This can be useful for multiple purposes but is also a security risk. Documentation should be updated to include security considerations of this particular setting so that Keycloak adopters can make the necessary security decisions. [KEYCLOAK-15947](https://issues.redhat.com/browse/KEYCLOAK-15947)
```

Reply:
Updated the wording.